### PR TITLE
Add NumberTextProvider

### PR DESCRIPTION
**Pull request by InfinityGhost**
# Changes
- Reduces some floating point errors caused by the built in `NumericMaskedTextBox<T>`

### CODE REVIEW

File: OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            {
+                builder.Clear();
+                if (value != null)
+				{
+					int pos = 0;
+					foreach (char ch in value)
+					{
+						Insert(ch, ref pos);
+					}
+				}
```
Comment (AbstractQbit) [inline OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs:29]: here

File: OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        public bool MaskCompleted => true;
+
+        public IEnumerable<int> EditPositions
+        {
+            get
+            {
+                for (int i = 0; i <= builder.Length; i++)
+				{
+					yield return i;
+				}
```
Comment (AbstractQbit) [inline OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs:43]: there

File: OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+			{
+				length = Math.Min(length, builder.Length - position);
+				builder.Remove(position, length);
+			}
+			else if (position >= length)
+			{
+				builder.Remove(position - length, length);
+				position = Math.Max(0, position - length);
+			}
+			return true;
```
Comment (AbstractQbit) [inline OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs:68]: also here

File: OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        public bool Replace(char character, ref int position)
+        {
+            if (Allow(ref character, ref position))
+            {
+                if (position >= builder.Length)
+                    return Insert(character, ref position);
+                builder[position] = character;
+                position++;
+            }
+			return true;
```
Comment (AbstractQbit) [inline OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs:91]: this

File: OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+						position++;
+				}
+				else
+					position++;
+				builder.Insert(0, character);
+				return false;
+			}
+			allow |= char.IsDigit(character);
+			return allow;
+		}
```
Comment (AbstractQbit) [inline OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs:142]: and everything there

**Review by AbstractQbit — CHANGES_REQUESTED** [review]
Mixing tab and space indents looks really ugly when viewed on gh, because tabs don't always get interpreted as 4 spaces, mb change tabs to spaces to match the rest of the code?

File: OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs (hunk 1 above)
Comment (InfinityGhost) [reply OpenTabletDriver.UX/Controls/Generic/Text/Providers/NumberTextProvider.cs:29]: lmao i stole this code from eto, will fix all of this

**Pull request merged**